chatsky/context_storages/file.py

```diff
@@ -0,0 +1,192 @@
+"""
+JSON
+----
+The JSON module provides a json-based version of the :py:class:`.DBContextStorage` class.
+This class is used to store and retrieve context data in a JSON. It allows Chatsky to easily
+store and retrieve context data.
+"""
+
+from abc import ABC, abstractmethod
+import asyncio
+from pickle import loads, dumps
+from shelve import DbfilenameShelf
+from typing import List, Set, Tuple, Dict, Optional
+import logging
+
+from pydantic import BaseModel, Field
+
+from .database import DBContextStorage, _SUBSCRIPT_DICT, _SUBSCRIPT_TYPE
+from chatsky.utils.logging import collapse_num_list
+
+try:
+    from aiofiles import open
+    from aiofiles.os import stat, makedirs
+    from aiofiles.ospath import isfile
+
+    json_available = True
+    pickle_available = True
+except ImportError:
+    json_available = False
+    pickle_available = False
+
+
+logger = logging.getLogger(__name__)
+
+
+class SerializableStorage(BaseModel):
+    main: Dict[str, Tuple[int, int, int, bytes, bytes]] = Field(default_factory=dict)
+    turns: List[Tuple[str, str, int, Optional[bytes]]] = Field(default_factory=list)
+
+
+class FileContextStorage(DBContextStorage, ABC):
+    """
+    Implements :py:class:`.DBContextStorage` with `json` as the storage format.
+
+    :param path: Target file URI. Example: `json://file.json`.
+    :param context_schema: Context schema for this storage.
+    :param serializer: Serializer that will be used for serializing contexts.
+    """
+
+    is_asynchronous = False
```

Comment:
What is the point of this flag?
Even if you limit concurrent execution within the processing of a single context, you still have potentially multiple contexts being processed at the same time.
IMO it's better to use `asyncio.lock` inside the methods that require to read and write data within a single method call.